joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h

```diff
@@ -246,8 +246,51 @@ class JointTrajectoryController : public controller_interface::Controller<Hardwa
    */
   void setHoldPosition(const ros::Time& time, RealtimeGoalHandlePtr gh=RealtimeGoalHandlePtr());
 
+  /**
+   * \brief Internal update function.
+   * Update time data, current states and errors and check tolerances.
+   */
+  virtual void updateStateAndTimeData(const ros::Time& time, const ros::Duration& period, TimeData& time_data);
+
+  /**
+   * \brief Update commands on the hardware interface.
+   */
+  virtual void updateHardwareInterface(const TimeData& time_data);
+
+  /**
+   * \brief Set current active goal to succeeded (if present).
+   * \note This method is realtime-safe.
+   */
+  virtual void setGoalAsSucceeded();
+
+  /**
+   * \brief Set action feedback (if active goal present).
+   * \note This method is realtime-safe.
+   */
+  virtual void setActionFeedback();
```

Comment:
I'm not a fan of all this `virtual` stuff because, in my opinion, it's violates the encapsulation. What is it you are trying to archive? Only expose the stuff, for what you actually have a usage.